cpp/dolfinx/refinement/uniform.cpp

```diff
@@ -0,0 +1,264 @@
+        for (std::size_t i = 0; i < nv_ent; ++i)
+        {
+          for (int k = 0; k < 3; ++k)
+            v[k] += x_g[3 * vertex_to_x[vt[i]] + k];
+        }
+        for (int k = 0; k < 3; ++k)
+          new_x[(w + w_off) * 3 + k] = v[k] / static_cast<T>(nv_ent);
+      }
+    }
+    common::Scatterer sc(*index_maps[j], 1);
+    sc.scatter_fwd(std::span<const std::int64_t>(new_v[j]),
+                   std::span(std::next(new_v[j].begin(), num_entities),
+                             index_maps[j]->num_ghosts()));
+    w_off += num_entities;
+  }
+
+  // Create new topology...
+  const std::vector<mesh::CellType>& cell_entity_types = entity_types.back();
+  std::vector<std::vector<std::int64_t>> mixed_topology(
+      cell_entity_types.size());
+
+  // Find index of tets in topology list, if any
+  int ktet = -1;
+  auto it = std::find(cell_entity_types.begin(), cell_entity_types.end(),
+                      mesh::CellType::tetrahedron);
+  if (it != cell_entity_types.end())
+    ktet = std::distance(cell_entity_types.begin(), it);
+  // Topology for tetrahedra which arise from pyramid subdivision
+  std::vector<int> pyr_to_tet_list
+      = {5, 13, 7, 9, 6, 13, 11, 7, 10, 13, 12, 11, 8, 13, 9, 12};
+
+  std::vector<int> refined_cell_list;
+  for (int k = 0; k < static_cast<int>(cell_entity_types.size()); ++k)
+  {
+    // Reserve an estimate of space for the topology of each type
+    mixed_topology[k].reserve(mesh.topology()->index_maps(tdim)[k]->size_local()
+                              * 8 * 6);
+
+    // Select correct subdivision for celltype
+    // Hex -> 8 hex, Prism -> 8 prism, Tet -> 8 tet, Pyr -> 5 pyr + 4 tet
+    if (cell_entity_types[k] == mesh::CellType::hexahedron)
+    {
+      spdlog::info("Hex subdivision [{}]", k);
+      refined_cell_list
+          = {0, 9,  8,  20, 10, 22, 21, 26, 1, 11, 8,  20, 12, 23, 21, 26,
+             2, 13, 9,  20, 14, 24, 22, 26, 3, 13, 11, 20, 15, 24, 23, 26,
+             4, 16, 10, 21, 17, 25, 22, 26, 5, 16, 12, 21, 18, 25, 23, 26,
+             6, 17, 14, 22, 19, 25, 24, 26, 7, 18, 15, 23, 19, 25, 24, 26};
+    }
+    else if (cell_entity_types[k] == mesh::CellType::tetrahedron)
+    {
+      spdlog::info("Tet subdivision [{}]", k);
+      refined_cell_list = {0, 7, 8, 9, 1, 5, 6, 9, 2, 4, 6, 8, 3, 4, 5, 7,
+                           9, 4, 6, 8, 9, 4, 8, 7, 9, 4, 7, 5, 9, 4, 5, 6};
+    }
+    else if (cell_entity_types[k] == mesh::CellType::prism)
+    {
+      spdlog::info("Prism subdivision [{}]", k);
+      refined_cell_list
+          = {0,  6,  7,  8,  15, 16, 6,  1,  9,  15, 10, 17, 7,  9,  2,  16,
+             17, 11, 6,  9,  7,  15, 17, 16, 15, 17, 16, 12, 14, 13, 8,  15,
+             16, 3,  12, 13, 11, 17, 16, 5,  14, 13, 10, 15, 17, 4,  12, 14};
+    }
+    else if (cell_entity_types[k] == mesh::CellType::pyramid)
+    {
+      spdlog::info("Pyramid subdivision [{}]", k);
+      refined_cell_list = {0,  5,  6, 13, 7,  1,  8,  5, 13, 9,  3,  10, 8,
```

Comment:
Could you explain what is in the list? I though it would be the local connectivity for the entires 
Above it i stated that a pyramid is subdivided in 5 pyramids and 4 tetrahedra. I would then expect the array to be (5*5 + 4*4) entries, while this list has only 25 entries. 
One can later observe that you have append the array: `pyr_to_tet_list` to mixed topology, which accounts for the missing tetrahedra.

This should be better documented.